include/NAS2D/StringUtils.h

```diff
@@ -54,5 +97,23 @@ namespace NAS2D
  */
 	using StringList = std::vector<std::string>;
 
+	namespace detail
+	{
+		template<typename T>
+		T rangeCheckHelper(const std::string& value)
+		{
+			const auto valueAsInteger = std::stoi(value);
+			if constexpr (std::numeric_limits<T>::max() < std::numeric_limits<int>::max())
+			{
+				if (std::numeric_limits<T>::max() < valueAsInteger)
+				{
+					throw std::out_of_range("rangeCheckHelper argument out of range");
+				}
+			}
+			return static_cast<T>(valueAsInteger);
+		}
+		extern template char rangeCheckHelper(const std::string& value);
+	}
+

```

Comment:
This is a private detail which is not used in the header file. It would be better off in an unnamed namespace in the implementation file.

Reply:
Is that even possible with templates? It was my understanding base templates are [typically defined in the header](https://www.youtube.com/watch?v=NIDEjY5ywqU&t=6m55s).

Reply:
Things (including templates) that are used by multiple translation units need to be in a header. This code is only used internally by one translation unit. It does not need to be in a header.

This is not to be confused with why template **definitions** are typically placed in headers, as opposed to only a **declaration**. That video is talking about definitions and usability (from multiple translation units). In that case, yes, the definition needs to be in the header, since templates need a full definition to be usable.

Aside: An exception to that, are templates with a specific set of type arguments that are declared to be `extern`. In that case, the compiler can assume the function will be defined elsewhere for that exact set of type parameters, and can rely on the linker to later find the right references. Note that a template with a full set of type parameters specified is basically equivalent to a regular function.